pkg/storage/local.go

```diff
@@ -0,0 +1,482 @@
+package storage
+
+import (
+	"context"
+	"fmt"
+	"io"
+	llog "log"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+
+	"github.com/TierMobility/boring-registry/pkg/core"
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+)
+
+type LocalFileSystem interface {
+	OpenFile(name string, flag int, perm os.FileMode) (io.ReadWriteCloser, error)
+	ReadFile(name string) ([]byte, error)
+	Stat(name string) (os.FileInfo, error)
+	ReadDir(name string) ([]os.DirEntry, error)
+	MkdirAll(name string, perm os.FileMode) error
+}
+
+type FileServer interface {
+	ListenAndServe() error
+	Addr() string
+}
+
+type LocalStorage struct {
+	fs                  LocalFileSystem
+	server              FileServer
+	serverEndpoint      string
+	storageDir          string
+	moduleArchiveFormat string
+}
+
+func NewLocalStorage(fs LocalFileSystem, server FileServer, storageDir, serverEndpoint, moduleArchiveFormat string) *LocalStorage {
+	// when use client to upload or migrate, no need to set up the http file server
+	if server != nil && !reflect.ValueOf(server).IsNil() {
+		go func() {
+			if err := server.ListenAndServe(); err != nil {
+				llog.Printf("error: %+v", err)
+			}
+		}()
+	}
+
+	return &LocalStorage{
+		fs:                  fs,
+		server:              server,
+		serverEndpoint:      serverEndpoint,
+		storageDir:          storageDir,
+		moduleArchiveFormat: moduleArchiveFormat,
+	}
+}
+
+func NewDefaultLocalStorage(storageDir, moduleArchiveFormat, endpoint string, serverAddr string) *LocalStorage {
+	var server *fileServer
+	if len(serverAddr) != 0 {
+		server = &fileServer{
+			addr: serverAddr,
+			path: storageDir,
+		}
+	}
```

Comment:
Personally, I feel like it's cleaner to instantiate the `fileServer` in the `setupStorage()` function in the `root.go` file.
The logic to check which command (upload, migrate,server) should also live there, which makes it easier to understand what exactly happens if the command `server` is passed in comparison to `upload` for example :)
Hope that makes sense